py/desitarget/htm.py

```diff
@@ -0,0 +1,393 @@
+       Each of the dictionary values has the same format as the input for vert (explained under 'Parameters')
+    """
+
+    childverts = {}
+    npix = len(vert)
+
+    #ADM Find the midpoint vectors of the parent triangle...
+    A = vert[:,1] + vert[:,2]
+    B = vert[:,0] + vert[:,2]
+    C = vert[:,0] + vert[:,1]
+    #ADM ...and normalize these vectors so that they're on the unit sphere
+    A /= np.linalg.norm(vert[:,1] + vert[:,2],axis=1,keepdims=True)
+    B /= np.linalg.norm(vert[:,0] + vert[:,2],axis=1,keepdims=True)
+    C /= np.linalg.norm(vert[:,0] + vert[:,1],axis=1,keepdims=True)
+    
+    #ADM Now compile test progeny triangles from the midpoints and vertices of the parent triangle
+    #ADM and put them in a dictionary where the key is the name of the node
+    #ADM The reshape and hstack-ing is just to format the triangles the same as the input format
+    childverts[0] = np.hstack(np.array([vert[:,0],C,B])).reshape(npix,3,3)
+    childverts[1] = np.hstack(np.array([vert[:,1],A,C])).reshape(npix,3,3)
+    childverts[2] = np.hstack(np.array([vert[:,2],B,A])).reshape(npix,3,3)
+    childverts[3] = np.hstack(np.array([A,B,C])).reshape(npix,3,3)
+
+    return childverts
+
+
+def lookup(ra,dec,level=20,charpix=True,verbose=True):
+    """Return the HTM pixel for a given RA/Dec
+
+    Parameters
+    ----------
+    ra : :class:`float` or `array`
+        A Right Ascension in degrees (can also be a numpy array with multiple values)
+    dec : :class:`float` or `array`
+        A Declination in degrees (can also be a numpy array with multiple values)
+    level : :class:`int`, optional
+        Which level of the HTM tree to pixelize down to
+    charpix : :class:`bool`, optional, defaults to True
+        If True, return pixels in character format, otherwise return them in integer format
+    verbose : :class:`bool`, optional, defaults to True
+        If True, print timing and warning messages to screen
+
+    Returns
+    -------
+    :class:`char array` or `int array`
+        The HTM pixels corresponding to the passed RA/Dec at the requisite level. Will be the same
+        length as length of ra and dec
+
+    Timing Notes
+    ------------
+    
+    Performs about 16000 level 20 (resolution ~ 0.4 arcsec) lookups per second on a NERSC login node
+    Performs about 25000 level 13 (resolution ~ 0.8 arcmin) lookups per second on a NERSC login node
+    Is about a factor of 3 slower for charpix = False
+
+    """
+
+    t0 = time()
+
+    #ADM convert input spherical coordinates to Cartesian
+    v = SkyCoord(ra=ra*u.degree, dec=dec*u.degree, frame='icrs')
+    v.representation = 'cartesian'
+
+    #ADM convert SkyCoord object to numpy array for speedier manipulation
+    #ADM .T is the transpose attribute
+    #ADM this conversion also allows either single floats or longer arrays to be passed
+    v = np.vstack(np.array([v.x.value,v.y.value,v.z.value])).T
```

Comment:
Is doing this transform via SkyCoord necessary?  Usually unitful astropy stuff is quite slow.  This reminded me of some code from @rainwoodman in a languishing PR in desimodel:
```python
def _embed_sphere(ra, dec):
    """ embed RA DEC to a uniform sphere in three-d """
    phi = np.radians(np.asarray(ra))
    theta = np.radians(90.0 - np.asarray(dec))
    r = np.sin(theta)
    x = r * np.cos(phi)
    y = r * np.sin(phi)
    z = np.cos(theta)
    return np.array((x, y, z)).T
```
With this, we could drop SkyCoord and astropy.units imports